packages/oft-evm/contracts/NativeOFTAdapter.sol

```diff
@@ -0,0 +1,107 @@
+ * @dev WARNING: The default NativeOFTAdapter implementation assumes LOSSLESS transfers, ie. 1 native in, 1 native out.
+ */
+abstract contract NativeOFTAdapter is OFTCore {
+
+    address public constant NATIVE_TOKEN_ADDRESS = address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE);
+
+    error InsufficientMessageValue(uint256 provided, uint256 required);
+    error CreditFailed();
+
+    /**
+     * @dev Constructor for the NativeOFTAdapter contract.
+     * @param _localDecimals The decimals of the native on the local chain (this chain). 18 on ETH.
+     * @param _lzEndpoint The LayerZero endpoint address.
+     * @param _delegate The delegate capable of making OApp configurations inside of the endpoint.
+     */
+    constructor(
+        uint8 _localDecimals,
+        address _lzEndpoint,
+        address _delegate
+    ) OFTCore(_localDecimals, _lzEndpoint, _delegate) {}
+
+    /**
+     * @dev Locks native sent by the sender as msg.value
+     * @param _amountLD The amount of native to send in local decimals.
+     * @param _minAmountLD The minimum amount to send in local decimals.
+     * @param _dstEid The destination chain ID.
+     * @return amountSentLD The amount sent in local decimals.
+     * @return amountReceivedLD The amount received in local decimals on the remote.
+     */
+    function _debit(
+        address /*_from*/,
+        uint256 _amountLD,
+        uint256 _minAmountLD,
+        uint32 _dstEid
+    ) internal virtual override returns (uint256 amountSentLD, uint256 amountReceivedLD) {
+        // @dev Lock native funds by moving them into this contract from the caller.
+        // No need to transfer here since msg.value moves funds to the contract.
+        (amountSentLD, amountReceivedLD) = _debitView(_amountLD, _minAmountLD, _dstEid);
+    }
+
+    /**
+     * @dev Credits native to the specified address.
+     * @param _to The address to credit the native to.
+     * @param _amountLD The amount of native to credit.
+     * @dev _srcEid The source chain ID.
+     * @return amountReceivedLD The amount of native ACTUALLY received.
+     */
+    function _credit(
+        address _to,
+        uint256 _amountLD,
+        uint32 /*_srcEid*/
+    ) internal virtual override returns (uint256 amountReceivedLD) {
+        // @dev Transfer tokens to the recipient.
+        (bool success, ) = payable(_to).call{value: _amountLD}("");
+        if (!success) {
+            revert CreditFailed();
+        }
+
+        // @dev In the case of NON-default NativeOFTAdapter, the amountLD MIGHT not be == amountReceivedLD.
+        return _amountLD;
+    }
+
+    function send(
+        SendParam calldata _sendParam,
+        MessagingFee calldata _fee,
+        address _refundAddress
+    ) public payable virtual override returns (MessagingReceipt memory msgReceipt, OFTReceipt memory oftReceipt) {
+        if (_fee.nativeFee + _sendParam.amountLD != msg.value) {
+            revert NotEnoughNative(msg.value);
+        }
```

Comment:
Found out while writing tests that `msg.value` must be the same as `_fee.nativeFee`, even though we are overriding `_payNative` to be empty. This is due to `EndpointV2Mock._assertMessagingFee` requiring `msg.value` to be equal to `_fee.nativeFee` [here](https://github.com/LayerZero-Labs/devtools/blob/c745cf8d0992e2f8df7d00388596fe1b9c4b160b/packages/test-devtools-evm-foundry/contracts/mocks/EndpointV2Mock.sol#L318-L323)

I will look for a way around this

Reply:
This actually may be due to the `msg.value` not being carried over to `EndpointV2Mock._assertMessagingFee`, still looking!

Reply:
Figured this out -- `_payNative` cannot be empty, it must just return the parameter